### DISCUSSION TITLE
LOWER not working for span matcher

### DISCUSSION THREAD

**Discussion by iamyihwa** [explosion/spaCy#11815] [Help: Coding & Implementations] [answered]
Applying 'LOWER' to span matcher does not seem to work well. 

brands = ['nike corp', 'adidas corp']
(1) patterns = [{"label": "BRAND", "pattern": brand} for brand in brands]

<img width="395" alt="image" src="https://user-images.githubusercontent.com/12494156/202188662-2def41b5-7716-4c84-9ec5-169e67c4135e.png">
(2) patterns = [{"label": "BRAND", "pattern": [{"LOWER": brand }]}   for brand in brands]
<img width="500" alt="image" src="https://user-images.githubusercontent.com/12494156/202188580-9b13048e-b0f9-4500-9e4f-ea9637968d28.png">

In scenarios (1) and (2) does not detect reliably the brand names. 

## How to reproduce the behaviour

```
brands = ['nike corp', 'adidas corp']
nlp = spacy.load("en_core_web_sm")
ruler = nlp.add_pipe("span_ruler")
#patterns = [{"label": "BRAND", "pattern": brand} for brand in brands]
patterns = [{"label": "BRAND", "pattern": [{"LOWER": brand }]}   for brand in brands]
with nlp.select_pipes(enable="tagger"):
    ruler.add_patterns(patterns)
print(patterns)

text = "nike corp is a brand."
doc1 = nlp(text)
print([(span.text, span.label_, span.start) for span in doc1.spans["ruler"]])

text = "NIKE CORP is a brand."
doc1 = nlp(text)
print([(span.text, span.label_, span.start) for span in doc1.spans["ruler"]])
```
<img width="523" alt="image" src="https://user-images.githubusercontent.com/12494156/202187157-df28f7e6-af49-4e74-8e7c-5a4b82b63c84.png">

<img width="392" alt="image" src="https://user-images.githubusercontent.com/12494156/202187370-977f781e-7a59-4e98-8d4f-b3f1ab06c3d3.png">

## Your Environment

* Operating System: Windows 
* Python Version Used: 3.7.2
* spaCy Version Used: 3.4.1 
* Environment Information:

**Comment by rmitsch** [comment, marked as answer]
Hi @iamyihwa, the components of the patterns passed on `span_ruler` have to [correspond to spaCy's tokenization](https://spacy.io/usage/rule-based-matching#adding-patterns). I.e. each part of your pattern has to align with a token as recognized by spaCy. 

If you want to want to match for spans consisting of multiple tokens (such as "nike corp" or "adidas corp"), the pattern has to reflect this (see [here](https://spacy.io/api/spanruler#add_patterns) for another example with a pattern for "san francisco"). So instead of 
```python
patterns = [{"label": "BRAND", "pattern": [{"lower": "nike corp"}]}]
```
you'll want to do:
```python
patterns = [{"label": "BRAND", "pattern": [{"lower": "nike", "lower": "corp"}]}]
```

In your example you can replace
```python
patterns = [{"label": "BRAND", "pattern": [{"LOWER": brand }]}   for brand in brands]
```
with 
```python
patterns = [{"label": "BRAND", "pattern": [{"lower": token} for token in brand.split()]} for brand in brands]
```
to obtain correct results.

**Reply by iamyihwa** [reply]
Thanks @rmitsch for your help! 
```
patterns = [{"label": "BRAND", "pattern": [{"lower": token} for token in brand.split()]} for brand in brands]
```
works perfect!

**Reply by iamyihwa** [reply]
Just was wondering how different ways to do [Rule Based Matchings](https://spacy.io/usage/rule-based-matching) are different in their uses. 
 (e.g.  'EntityRuler', 'SpanMatcher', 'PhraseMatcher', 'TokenMatcher', etc.) are very different. 
I intended to do 'EntityRuler' using 'SpanMatcher'. 
In fact the solution is by using 'EntityRuler' with 'TokenMatcher'..

**Reply by rmitsch** [reply]
An alternative way to achieve the same goal would be to configure your `EntityRuler` or `SpanRuler` like so (see the [docs](https://spacy.io/api/entityruler#config)):
```python
ruler = nlp.add_pipe("span_ruler", config={"phrase_matcher_attr": "lower"})
```
Then you can add your patterns without breaking them down into tokens:
```python
patterns = [{"label": "BRAND", "pattern": brand} for brand in brands]
```

There are [some caveats](https://github.com/explosion/spaCy/issues/6994) regarding the tokenization though that might or might not be relevant for you if you choose to use this approach.

**Reply by iamyihwa** [reply]
Thank you for your explanation @rmitsch ! 
Sure will use the first approach but will have the second one also in mind! 

Thanks for the great work and help!!!
